Add endpoint to get SAML providers for a user.

jcdyer opened the pull request:
Adds an endpoint to allow mobile apps to get SSO providers for a given user.  Includes throttling to prevent the endpoint from being used to crawl for logins.  Due to policy concerns around potential leakage of user info, this is hidden behind a feature flag.

Upstreamed from edx-solutions/edx-platform#1047 and edx-solutions/edx-platform#1080

**JIRA tickets**: Implements OC-4285.

**Dependencies**: None

**Screenshots**:

**Sandbox URL**: https://pr18237.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

Set ALLOW_UNPRIVILEGED_SSO_PROVIDER_QUERY to true in lms.env.json.

Verify that the SSO user endpoint can be accessed by non-logged in users. Example: https://pr18237.sandbox.opencraft.hosting/api/third_party_auth/v0/users/staff

Verify that provider information can be retrieved by providing an email address. Example: https://pr18237.sandbox.opencraft.hosting/api/third_party_auth/v0/users/staff@example.com

**Author notes and concerns**:

**Reviewers**
- [x] @xitij2000 
- [ ] edX reviewer[s] TBD

**Settings**
Set ALLOW_UNPRIVILEGED_SSO_PROVIDER_QUERY to true in lms.env.json.

edx-webhook commented:
Thanks for the pull request, @jcdyer! I've created [OSPR-2428](https://openedx.atlassian.net/browse/OSPR-2428) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

mduboseedx commented:
@jcdyer Feel free to ping me once this passes all the tests. Thanks!

jcdyer commented:
@mduboseedx Thanks.  I'm unable to find the diff quality report on the jenkins tests.  Have those been removed or just relocated?  It'd difficult to determine which issue is causing the quality build to fail without that information.

jcdyer commented:
I think I found the source of the errors, (.coffee files that shouldn't have ben included), but I'm not sure how to diagnose that from the jenkins reports anymore.

mduboseedx commented:
@edx/testeng Do you know why these quality test is failing? We're having trouble finding the diff quality report.

jmbowman commented:
@mduboseedx It's the "diff_quality_pylint.html" artifact from the `edx-platform-quality-diff` sub-job: https://build.testeng.edx.org/job/edx-platform-quality-diff/2603/artifact/reports/diff_quality/diff_quality_pylint.html .  Two unused imports.

jcdyer commented:
Thanks @jmbowman.

mduboseedx commented:
@edx/platform-team Could you give this a review?

jcdyer commented:
@jmbowman Can you give this a review anytime soon, or find someone to do so?

macdiesel commented:
@jcdyer This is on my radar now. I gave it a quick once over and it looks fine. If you can get tests passing I'll take a closer look and merge.

jcdyer commented:
jenkins run bokchoy

mduboseedx commented:
jenkins run bokchoy

mduboseedx commented:
@jcdyer Have you had a chance to look into the failed bokchoy test?

doctoryes commented:
jenkins run bokchoy

xitij2000 commented:
jenkins run bokchoy

jcdyer commented:
Thanks @doctoryes.  I should be able to look into this again the week after next.  I'll ping you when it's ready for another review.

jcdyer commented:
@doctoryes I just rebased onto the latest master.  Whatever the issue was, it seems to be resolved now.

macdiesel commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -22,13 +23,42 @@
+    """
+    Maximum number of provider requests in a quick burst.
+    """
+    rate = '10/min'
```
Why not make this and the ProviderSustainedThrottle configuration settings?

macdiesel commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -68,26 +99,22 @@ def get(self, request, username):
+            if '@' in username:
+                user = User.objects.get(email=username)
+            else:
+                user = User.objects.get(username=username)
```
@pwnage101 Don't we have a method that does this?

pwnage101 replied:
yes, use `from student.models import get_user_by_username_or_email`

pwnage101 commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -22,13 +23,42 @@
     **Example Request**
 
         GET /api/third_party_auth/v0/users/{username}
+        GET /api/third_party_auth/v0/users/{email@example.com}
```
Can you explain why the username identifier is made into an ambiguous identifier here?  In general, I think we were trying to phase out ambiguous identifiers whenever possible, so I have to ask when we're adding new ones.

jcdyer replied:
Phase it out in favor of what?  

Our client that we did the downstream work for had a requirement of allowing users to log in by either identifier. 

Would you prefer to have different endpoints for each, or to use a query parameter to identify the type of identifier?

jcdyer replied:
Thanks @pwnage101.  That's helpful information.

FYI @xitij2000 and @bradenmacdonald: This is related to the SSO provider lookup we did recently.  This would require a minor change and the client side (having the api client choose an identifier type before making the request) and the server side (providing the argument as a query parameter) in the downstream code.  Can we do that work to help edx move forward with unambiguous user identifiers?

bradenmacdonald replied:
We can't immediately change the API contract, because it would break the existing apps, but if we did want to push the "username vs. email" logic into the clients and make this API optionally unambiguous (by requiring a parameter/path to explicitly specify email vs. username), and later require it to be unambiguous, that would be fine.

doctoryes replied:
I'll chime in here in summary: For this PR to proceed, we'd very much like it to include functionality to allow clients to unambiguously state whether they're sending an email address or a username, whether it's by two different new URLs -or- optional parameters on this endpoint. The old ambiguous endpoint would remain functional for now until clients can be shifted to use the unambiguous endpoint. But there'd be no remaining blocker to shifting clients over to use the unambiguous endpoint.

After clients are shifted, we'd then remove the old ambiguous endpoint functionality - perhaps in a new API version, if needed. Thoughts?

jcdyer replied:
Sounds good to me.  Here's how I'll implement that:

The proposed endpoint will remain.  A new route will be added at `/api/third_party_auth/v0/users/` (with no path segment for the ambiguous identifier), that takes one of `?username=` or `?email=`.  If neither is provided, the endpoint will return a 400.

jcdyer commented:
I will resolve review suggestions tonight or tomorrow morning.

jcdyer commented:
@macdiesel, @pwnage101: Sorry for the delay.  This is ready for another review now.  I have added a new view with an explicit username / email query parameter, made throttles configurable via settings, and added tests for throttling.

pwnage101 commented on common/djangoapps/third_party_auth/api/urls.py:
```diff
@@ -14,6 +14,7 @@
         UserView.as_view(),
```
I vaguely recall even wrapping this deprecated url pattern behind a feature toggle/django setting, but that may have been part of conversations with my team.  @doctoryes @macdiesel what are your thoughts about making this togglable?

pwnage101 replied:
Oh, well we can't do that of course because it impacts an active endpoint.  I'll just drop this concern for now.

pwnage101 commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -20,13 +24,129 @@
     """
     List the third party auth accounts linked to the specified user account.
 
+    [DEPRECATED]
```
nit: add a sentence about renaming UserView2 to UserView after ripping this one out.

jcdyer replied:
Better yet, I'll rename now, and call the old one UserViewDeprecated

jcdyer replied:
On second thought, there could be extension libraries that are referring to this.  Safer just to add the note.

bmedx commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -20,13 +24,129 @@
+
+        Params must be a dict that includes only one of 'username' or 'email'
+        """
+        assert identifier.kind in {'username', 'email'}
```
Any reason not to use `self.identifier_kinds` here? Could you also change this to raise a ValueError instead of using an `assert`?

bmedx replied:
@jcdyer I'm still seeing a conflict here. Can you take one more look?

jcdyer replied:
You are correct.  The rebase didn't take properly.  Sorry about that!

jcdyer commented:
Sorry.  I rebased against the wrong branch, and then pushing didn't work.  I'm resolving conflicts now.

edx-webhook commented:
@jcdyer 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18237&entry.1484655318=https://github.com/jcdyer&entry.752974735=2018-05-20+17:52&entry.1917517419=2018-10-18+18:38&entry.2133058324=Yes) so we can improve your experience in the future.

bmedx commented:
@jcdyer 🚢 thanks for your patience getting this through!

pwnage101 commented on common/djangoapps/third_party_auth/api/views.py:
```diff
@@ -66,34 +182,80 @@ def get(self, request, username):
-            }
-            for assoc in providers if assoc.has_account
-        ]
+# TODO: When removing deprecated UserView, rename this view to UserView.
```
Do we have an schedule estimate for addressing this tech debt?

jcdyer replied:
I think our client would want to have at least one named release that includes the backwards-compatible code, and it could be removed any time after that.
